datafusion/physical-plan/src/execution_plan.rs

```diff
@@ -885,7 +886,13 @@ pub fn check_not_null_contraits(
             );
         }
 
-        if batch.column(index).null_count() > 0 {
+        if batch
+            .column(index)
+            .logical_nulls()
+            .map(|nulls| nulls.null_count())
+            .unwrap_or_default()
+            > 0
```

Comment:
This could be simplified back if we had something like https://github.com/apache/arrow-rs/pull/6608